### PR TITLE
Add YARD documentation comments

### PR DESCRIPTION
**Pull request by roback**
Documented all public methods/attributes using http://yardoc.org/

Started a yard server locally by running

    bundle exec yard server --reload

### Todo

- [x] Rake task for generating documentation/starting server
- [x] Update readme
- [x] Answer the question: "should documentation be checked in or not?"

### CODE REVIEW

**Comment by roback** [comment]
> should documentation be checked in or not?

I have no idea, some keeps it in the repo, some don't. 

Im choosing not to as it's one more thing to remember to update each time you make a change. The YARD-comments in the code is still there which should be enough.

**Comment by dentarg** [comment]
I agree, we should not have the generated documentation in the repo

**Comment by dentarg** [comment]
:+1:

**Comment by jage** [comment]
Great job! :+1:

File: lib/twingly/search/client.rb [hunk 1]
```diff
@@ -10,20 +11,39 @@ class Client
...
       DEFAULT_USER_AGENT = "Twingly Search Ruby Client/#{VERSION}"
 
+      # Creates a new Twingly Search API client
+      #
+      # @param api_key [optional, String] the API key provided by Twingly.
+      #   If nil, reads api_key from environment (TWINGLY_SEARCH_KEY).
+      # @param options [Hash]
+      # @option options [String] :user_agent the user agent to be used
+      #    for all requests
+      # @raise [AuthError] if an API key isn't set.
```
Comment (walro) [inline lib/twingly/search/client.rb]: I think we can "is not" over "isn't" in our documentation, wdyt?

File: lib/twingly/search/error.rb [hunk 2]
```diff
@@ -1,6 +1,9 @@
 module Twingly
   module Search
     class Error < StandardError
+      # @param [String] message API response error message
+      # @return [Error] an instance of an {AuthError} or {ServerError},
+      #   which one depends on the message.
```
Comment (walro) [inline lib/twingly/search/error.rb]: > which one depends on the message

Thinking we don't need to document that

File: lib/twingly/search/parser.rb [hunk 3]
```diff
@@ -3,6 +3,11 @@
 module Twingly
   module Search
     class Parser
+      # Parse an API response body.
+      #
+      # @param [String] document containing an API response XML.
+      # @raise [Error] which error depends on the API response (see {Error.from_api_response_message}).
+      # @return [Result] containing the result for this.
```
Comment (walro) [inline lib/twingly/search/parser.rb]: > for this.

What's "this" in this case? I think we can remove "for this".

File: lib/twingly/search/post.rb [hunk 4]
```diff
@@ -4,10 +4,29 @@
...
+    #   language that the post was written in.
+    # @attr_reader [DateTime] indexed the time, in UTC, when this post was indexed by Twingly.
+    # @attr_reader [DateTime] published the time, in UTC, when this post was published.
+    # @attr_reader [String] blog_url the blog URL.
+    # @attr_reader [String] blog_name name of the blog.
+    # @attr_reader [String] authority the blog's authority/influence.
+    #   See https://developer.twingly.com/resources/search/#authority
+    # @attr_reader [Integer] blog_rank the rank of the blog, based on authority and language.
+    #   See https://developer.twingly.com/resources/search/#authority
+    # @attr_reader [Integer] tags
```
Comment (walro) [inline lib/twingly/search/post.rb]: are tags an integer?

File: lib/twingly/search/query.rb [hunk 5]
```diff
@@ -2,25 +2,44 @@
 
 module Twingly
   module Search
+    # Twingly Search API query
+    #
+    # @attr [String] pattern the search query.
+    # @attr [String] language which language restrict the query to.
```
Comment (walro) [inline lib/twingly/search/query.rb]: missing a "to" here

File: lib/twingly/search/query.rb [hunk 6]
```diff
@@ -2,25 +2,44 @@
...
+    # @attr [Client] client the client that this query is connected to.
+    # @attr [#to_time] start_time search for posts published after this time.
+    # @attr [#to_time] end_time search for posts published after this time.
```
Comment (walro) [inline lib/twingly/search/query.rb]: are these inclusive or exclusive? I don't think it should say "to_time" on both either

**Comment by walro** [comment]
Awesome work, left a few minor comments

File: lib/twingly/search/client.rb (hunk 1 above)
Comment (roback) [inline lib/twingly/search/client.rb]: Yes, that looks better.

File: lib/twingly/search/parser.rb (hunk 3 above)
Comment (roback) [inline lib/twingly/search/parser.rb]: > What's "this" in this case?

I don't remember, feels like something is missing at the end of the sentence. I'll just remove it :)

File: lib/twingly/search/post.rb (hunk 4 above)
Comment (roback) [inline lib/twingly/search/post.rb]: They are not :)

File: lib/twingly/search/query.rb (hunk 6 above)
Comment (roback) [inline lib/twingly/search/query.rb]: > I don't think it should say #to_time on both either

Yes it should (http://www.rubydoc.info/gems/yard/file/docs/Tags.md#Duck-Types)

Ill add `Time` too as that is what we use in our examples.

**Comment by roback** [comment]
> Awesome work, left a few minor comments

"a few" :stuck_out_tongue:

File: lib/twingly/search/query.rb (hunk 6 above)
Comment (walro) [inline lib/twingly/search/query.rb]: Ah, to_time means we will use it as a Time object? I thought of it as to/from.

**Comment by walro** [comment]
:) LGTM!

**Pull request merged**